[4.0] Convert plugin system privacyconsent to prepared statements

HLeithner opened the pull request:
# Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.


### Testing Instructions
Use the plugin in all ways you can think of.

Create a privacy consent form and test if it works.


### Expected result
Nothing changed.

HLeithner committed:
Convert plugin system privacyconsent to prepared statements

Quy commented on plugins/system/privacyconsent/privacyconsent.php:
```diff
@@ -553,13 +564,14 @@ private function remindExpiringConsents()
-			->where($db->quoteName('remind') . ' = 0');
-		$query->where($query->dateAdd($db->quote($now), $period, 'DAY') . ' > ' . $db->quoteName('created'));
+			->where($db->quoteName('remind') . ' = 0')
+			->where($query->dateAdd(':now', $period, 'DAY') . ' > ' . $db->quoteName('created'))
```
Bind `$period` like here? https://github.com/joomla/joomla-cms/pull/25503/files#diff-bbf838ec8885d4ea43274f7b94e17808R302

Quy commented on plugins/system/privacyconsent/privacyconsent.php:
```diff
@@ -655,11 +670,12 @@ private function invalidateExpiredConsents()
+		$query->select($db->quoteName(['id', 'user_id']))
 			->from($db->quoteName('#__privacy_consents'))
-			->where($query->dateAdd($db->quote($now), $period, 'DAY') . ' > ' . $db->quoteName('created'))
+			->where($query->dateAdd(':now', $period, 'DAY') . ' > ' . $db->quoteName('created'))
```
Bind `$period` like here? https://github.com/joomla/joomla-cms/pull/25503/files#diff-bbf838ec8885d4ea43274f7b94e17808R302

Quy commented on plugins/system/privacyconsent/privacyconsent.php:
```diff
@@ -185,10 +187,10 @@ public function onUserAfterSave($data, $isNew, $result, $error)
 
 		if ($option == 'com_users'
-			&&in_array($task, array('registration.register', 'profile.save'))
+			&&in_array($task, ['registration.register', 'profile.save'])
```
```suggestion
			&& in_array($task, ['registration.register', 'profile.save'])
```

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

HLeithner committed:
Add missing prepared parameter

Quy commented on plugins/system/privacyconsent/privacyconsent.php:
```diff
@@ -685,10 +702,12 @@ private function invalidateExpiredConsents()
-				->where($db->quoteName('id') . ' = ' . (int) $user->id);
+				->set($db->quoteName('state') . ' = 0')
+				->where($db->quoteName('id') . ' = :userid')
+				->bind(':userid', $userId);
```
```suggestion
				->bind(':userid', $userId, ParameterType::INTEGER);
```

HLeithner committed:
Update plugins/system/privacyconsent/privacyconsent.php

Co-Authored-By: Quy <quy@fluxbb.org>

HLeithner committed:
Split core.js and remove webcomponent

HLeithner committed:
Add webcomponent loader original copy from webcomponents/polyfills

https://github.com/webcomponents/polyfills/blob/14c7b973762c22e9b8141dcffe530219a8ba3cbc/packages/webcomponentsjs/webcomponents-loader.js

HLeithner committed:
Convert webcomponent to es6 and integrate into Joomla API preserving copyright.

Based on https://github.com/joomla/joomla-cms/blob/470851b1c7020a9b5bb212dc0328d1600381653a/build/media_source/system/js/core.es6.js

HLeithner committed:
Merge branch '4.0-dev' of github.com:HLeithner/joomla-cms into 4.0-dev

SharkyKZ commented on plugins/system/privacyconsent/privacyconsent.php:
```diff
@@ -551,15 +562,16 @@ private function remindExpiringConsents()
-		$query->where($query->dateAdd($db->quote($now), $period, 'DAY') . ' > ' . $db->quoteName('created'));
+			->where($db->quoteName('remind') . ' = 0')
+			->where($query->dateAdd(':now', ':period', 'DAY') . ' > ' . $db->quoteName('created'))
+			->bind([':now', ':period'], [$now, $period], [ParameterType::STRING, ParameterType::INTEGER]);
```
Use separate `bind()` calls or assign `[$now, $period]` to a variable.

And this is failing on PostgreSQL:

    SQLSTATE[HY093]: Invalid parameter number: :period

HLeithner replied:
Fixed thanks

HLeithner committed:
Merge remote-tracking branch 'origin/4.0-dev' into prepared-plg-system-privacyconsent

HLeithner committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

# Conflicts:
#	plugins/system/privacyconsent/privacyconsent.php

HLeithner committed:
Revert "Convert webcomponent to es6 and integrate into Joomla API preserving copyright."

This reverts commit 109432b

HLeithner committed:
Revert "Add webcomponent loader original copy from webcomponents/polyfills"

This reverts commit 7a1e963

HLeithner committed:
Revert "Split core.js and remove webcomponent"

This reverts commit 3e40e54

HLeithner committed:
Fix reference

HLeithner committed:
cs

wilsonge committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

waader commented:
I have tested this item :red_circle: unsuccessfully on df9ea6f3de6cf9c1f2508ff221abf876c57e44b4<br /><br />I logged in on frontend an when submitting the profil page with the privacy consent I get the error: 0 Class 'ArrayHelper' not found

Tested with postgresql.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25540">issues.joomla.org/tracker/joomla-cms/25540</a>.</sub>

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

richard67 commented on plugins/system/privacyconsent/privacyconsent.php:
```diff
@@ -15,21 +15,22 @@
 use Joomla\Component\Actionlogs\Administrator\Model\ActionlogModel;
 use Joomla\Component\Messages\Administrator\Model\MessageModel;
 use Joomla\Database\Exception\ExecutionFailureException;
-use Joomla\Utilities\ArrayHelper;
```
This line "use Joomla\Utilities\ArrayHelper;" should not be removed because the array helper is still used when this PR is applied. It explains waader's bad test.

richard67 replied:
E.g. in routine "onUserBeforeSave": "$userId = ArrayHelper::getValue($user, 'id', 0, 'int');".

HLeithner replied:
ArrayHelper added and Table removed because not used in this file.

HLeithner committed:
Fix imports

richard67 commented:
@waader Good find. Error has been corrected here meanwhile. Could you test again? Thanks in advance.

richard67 commented:
I have tested this item :white_check_mark: successfully on 72a763b7218684907e8993ae03f983fbfded1e5b<br /><br />Hint for other testers: Untranslated language strings are not related to or subject of this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25540">issues.joomla.org/tracker/joomla-cms/25540</a>.</sub>

waader commented:
I have tested this item :white_check_mark: successfully on 72a763b7218684907e8993ae03f983fbfded1e5b<br /><br />The error has gone!

I noticed that the request for information deletion as well as export is send twice to the user. This has probably nothing to do with the work done here?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25540">issues.joomla.org/tracker/joomla-cms/25540</a>.</sub>

richard67 commented:
@waader I don't think it is related to this PR.

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25540">issues.joomla.org/tracker/joomla-cms/25540</a>.</sub>

Commit:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

wilsonge commented:
Thanks!

wilsonge commented:
@waader can you send a separate issue in for the double email sending please?
